File: scripts/fetch-fbref-schedule.py
"""
Fetch FBref schedule using soccerdata (bypasses Cloudflare via undetected-chromedriver).

Requires Chrome/Chromium installed. Set path via env var SOCCERDATA_BROWSER
or the script auto-detects Playwright's bundled Chrome.
"""

from json import dumps
from os.path import dirname, join
from sys import exit as sysexit

from utils import YEAR, resolve_datetime, maybe_get_chrome_path

SCORE_DELIMITER = "\u2013"
OUTPUT_PATH = join(dirname(__file__), "references", f"{YEAR}-fbref-schedule.json")


def parse_score(score_str: str):
    if not score_str or score_str == SCORE_DELIMITER:
        return None
    parts = score_str.split(SCORE_DELIMITER)
    if len(parts) != 2:
        return None
    try:
        return {"home": int(parts[0]), "away": int(parts[1])}
    except ValueError:
        return None


def main():
    from soccerdata import FBref

    chrome_path = maybe_get_chrome_path()
    if not chrome_path:
        print(
            "Chrome not found. Install Chrome or set SOCCERDATA_BROWSER env var "
            "pointing to your Chrome/Chromium binary."
        )
        sysexit(1)

    print(f"Using Chrome: {chrome_path}")

    fbref = FBref(
        leagues="ENG-Premier League",
        seasons=YEAR,
        path_to_browser=chrome_path,
        headless=True,
    )

    print(f"Fetching FBref schedule for {YEAR}...")
    df = fbref.read_schedule()

    matches = []
    for _, row in df.iterrows():
        score_raw = row.get("score") or ""
        score = parse_score(score_raw)

        attendance = row.get("attendance")
        if attendance is not None and attendance != attendance:
            attendance = None
        if attendance is not None:
            try:
                attendance = int(attendance)
            except (TypeError, ValueError):
                attendance = None

        match_report_url = row.get("match_report")
        if match_report_url and not match_report_url.startswith("http"):
            match_report_url = "https://fbref.com" + match_report_url

        date_val = row.get("date")
        if hasattr(date_val, "strftime"):
            date_val = date_val.strftime("%Y-%m-%d")

        time_str = row.get("time") or ""
        matches.append({
            "matchweek": int(row["week"]),
            "datetime": resolve_datetime(date_val, time_str),
            "home": row["home_team"],
            "away": row["away_team"],
            "score": score,
            "attendance": attendance,
            "venue": row.get("venue") or "",
            "referee": row.get("referee") or None,
            "matchReportUrl": match_report_url,
        })

    print(f"  Parsed {len(matches)} matches for {YEAR}")

    with open(OUTPUT_PATH, "w") as f:
        f.write(dumps(matches, indent=2, ensure_ascii=False))

    print(f"Saved {OUTPUT_PATH}")


if __name__ == "__main__":
    try:
        main()
    except Exception as error:
        print("Fatal error:", error)
        sysexit(1)

File: scripts/fetch-fbref-schedule.ts
import axios from 'axios'
import fs from 'fs/promises'
import { JSDOM } from 'jsdom'
import type { FbrefScheduleEntry } from './types'

const YEARS = [2023, 2024, 2025]

const FBREF_SEASON_MAP: Record<string, string> = {
  '2023': '2023-2024',
  '2024': '2024-2025',
  '2025': '2025-2026',
}

function parseScore(scoreStr: string): { home: number; away: number } | null {
  if (!scoreStr || scoreStr === '–') return null
  const parts = scoreStr.split('–')
  if (parts.length !== 2) return null
  return { home: parseInt(parts[0], 10), away: parseInt(parts[1], 10) }
}

function parseAttendance(attStr: string): number | null {
  if (!attStr) return null
  return parseInt(attStr.replace(/,/g, ''), 10) || null
}

async function fetchSchedule(year: number): Promise<FbrefScheduleEntry[]> {
  const seasonPath = FBREF_SEASON_MAP[year]
  const url = `https://fbref.com/en/comps/9/${seasonPath}/schedule/${seasonPath}-Premier-League-Scores-and-Fixtures`

  console.log(`Fetching FBref schedule for ${year}...`)
  const response = await axios.get(url, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
      Accept: 'text/html',
    },
    timeout: 20000,
  })

  const dom = new JSDOM(response.data)
  const doc = dom.window.document

  const table = doc.querySelector(`table[id^="sched_${seasonPath}"]`)
  if (!table) {
    throw new Error(`Schedule table not found for ${year}`)
  }

  const rows = table.querySelectorAll('tbody tr')
  const matches: FbrefScheduleEntry[] = []

  for (const row of rows) {
    const cells = row.querySelectorAll('td, th')
    if (cells.length < 11) continue

    const matchweek = parseInt(cells[0].textContent!.trim(), 10)
    if (isNaN(matchweek)) continue

    const date = cells[2].textContent!.trim()
    const time = cells[3].textContent!.trim()
    const home = cells[4].textContent!.trim()
    const scoreRaw = cells[5].textContent!.trim()
    const away = cells[6].textContent!.trim()
    const attendanceRaw = cells[7].textContent!.trim()
    const venue = cells[8].textContent!.trim()
    const referee = cells[9].textContent!.trim()

    const matchReportLink = cells[10].querySelector('a')
    const matchReportUrl = matchReportLink ? matchReportLink.getAttribute('href') : null

    const score = parseScore(scoreRaw)
    const attendance = parseAttendance(attendanceRaw)

    matches.push({
      matchweek,
      date,
      time,
      home,
      away,
      score,
      attendance,
      venue,
      referee: referee || null,
      matchReportUrl: matchReportUrl ? `https://fbref.com${matchReportUrl}` : null,
    })
  }

  console.log(`  Parsed ${matches.length} matches for ${year}`)
  return matches
}

async function main() {
  for (const year of YEARS) {
    const matches = await fetchSchedule(year)
    const outputPath = `scripts/references/${year}-fbref-schedule.json`
    await fs.writeFile(outputPath, JSON.stringify(matches, null, 2))
    console.log(`Saved ${outputPath}`)
  }
}

main().catch((error: unknown) => {
  console.error('Fatal error:', error)
  process.exit(1)
})
